src/doc/rustc/src/platform-support/wasm32-unknown-emscripten.md

```diff
@@ -0,0 +1,116 @@
+suited for you. However, [`wasm32-unknown-unknown`](./wasm32-unknown-unknown.md)
+only has partial support for the Rust standard library and does not (easily)
+support interop with C/C++ code.
+
+Like Emscripten, the WASI targets [`wasm32-wasip1`](./wasm32-wasip1.md) and
+[`wasm32-wasip2`](./wasm32-wasip2.md) also provide access to the host environment,
+support interop with C/C++ (and other languages), and support most of the Rust
+standard library. While the WASI targets are portable across different hosts
+(web and non-web), WASI has no standard way of accessing web APIs, whereas
+Emscripten has the ability to run arbitrary JS from WASM and access many web APIs.
+If you are only targeting the web and need to access web APIs, the
+`wasm32-unknown-emscripten` target may be preferable.
+
+## Target maintainers
+
+- Hood Chatham, https://github.com/hoodmane
+- Sam Clegg, https://github.com/sbc100
+- Juniper Tyree, https://github.com/juntyr
+
+## Requirements
+
+This target is cross-compiled. The Emscripten compiler toolchain (emsdk) must be
+installed to link Rust-compiled code with Emscripten's libc and other C/C++
+code. Please refer to <https://emscripten.org/docs/getting_started/downloads.html>
+on how to get started with the emsdk.
+
+## Building the target
+
+Building this target can be done by:
+
+* Configure the `wasm32-unknown-emscripten` target to get built.
+* Configure LLD to be built.
+* Ensure the `WebAssembly` target backend is not disabled in LLVM.
+
+These are all controlled through `config.toml` options. It should be possible
+to build this target on any platform.
+
+## Building Rust programs
+
+Rust programs can be compiled by adding this target via rustup:
+
+```sh
+$ rustup target add wasm32-unknown-emscripten
+```
+
+and then compiling with the target:
+
+```sh
+$ rustc foo.rs --target wasm32-unknown-emscripten
+$ file foo.wasm
+```
+
+## Cross-compilation
+
+This target can be cross-compiled from any host.
+
+## Emscripten ABI Compatibility
+
+The Emscripten compiler toolchain does not follow a semantic versioning scheme
+that clearly indicates when breaking changes to the ABI can be made. If you are
+using the pre-compiled `std` from rustup, it may be compiled with a different
+Emscripten version than the local Emscripten you have installed. If a breaking
+change to the ABI was made in between the two versions, your code will exhibit
+undefined behaviour. If you come across issues, you can rebuild the Rust standard
+library with your local Emscripten version using
+
+```sh
+$ cargo +nightly -Zbuild-std build
+```
+
+## Testing
+
+This target is not extensively tested in CI for the rust-lang/rust repository.
+It's recommended to test the [`wasm32-wasip1`](./wasm32-wasip1.md) target instead
+for WebAssembly compatibility.
+
+## Conditionally compiling code
+
+It's recommended to conditionally compile code for this target with:
+
+```text
+#[cfg(all(target_family = "wasm", target_os = "emscripten"))]
```

Comment:
I'd say it's sufficient to do:
```
#[cfg(target_os = "emscripten")]
```
It's often necessary to filter for wasm, not emscripten somewhere else:
```
#[cfg(all(target_family = "wasm", not(target_os = "emscripten)))]` 
```
I've hit a bunch of duplicate definition compile errors, though if people are building for Emscripten at all they would catch this one.

In a lot of cases the Emscripten branch is the same as the linux branch since Emscripten implements a lot of linux headers.

Reply:
I've updated the text - thanks!